crates/solvers/src/domain/solution.rs

```diff
@@ -34,33 +34,43 @@ pub enum Trade {
 pub struct Fulfillment {
     order: order::Order,
     executed: U256,
+    fee: Fee,
 }
 
 impl Fulfillment {
     /// Creates a new order filled to the specified amount. Returns `None` if
     /// the fill amount is incompatible with the order.
-    pub fn partial(order: order::Order, executed: U256) -> Option<Self> {
-        let fill = match order.side {
-            order::Side::Buy => order.buy.amount,
-            order::Side::Sell => order.sell.amount,
-        };
+    pub fn new(order: order::Order, executed: U256, fee: Fee) -> Option<Self> {
+        if matches!(fee, Fee::Surplus(_)) != order.has_solver_fee() {
+            return None;
+        }
 
-        if (!order.partially_fillable && executed != fill)
-            || (order.partially_fillable && executed > fill)
+        let (fill, full) = match order.side {
+            order::Side::Buy => (order.buy.amount, executed),
+            order::Side::Sell => (
+                order.sell.amount,
+                executed.checked_add(fee.surplus().unwrap_or_default())?,
+            ),
+        };
+        if (!order.partially_fillable && full != fill) || (order.partially_fillable && full > fill)
```

Comment:
Shouldn't this rather say `order.partially_fillable && fill > full` to make sense? But I think it works out because `fill` and `full` got mixed up during initialization.